lib/bemhtml/compiler.js

```diff
@@ -375,14 +397,57 @@ Compiler.prototype.flatten = function flatten(expr) {
         callee: { type: 'Identifier', name: 'template' },
         arguments: t.predicates
       },
-      arguments: [ this.replaceBody(t.body) ]
+      arguments: [ this.replaceBody(t.body, extendedModeDetected) ]
     };
   }, this);
 };
 
-Compiler.prototype.replaceBody = function replaceBody(body) {
+Compiler.prototype.replaceBody = function replaceBody(body, extendedModeDetected) {
   var self = this;
 
+  // Desugar bodies of replace() and extend()
+  if (extendedModeDetected) {
+    body.body = estraverse.replace(body.body, {
+      enter: function (ast, parent) {
+        // don't mess with nested scopes
+        if (ast.type === 'FunctionDeclaration' ||
+            ast.type === 'FunctionExpression') {
+          this.skip();
+        }
+      },
+      leave: function (ast) {
+        if (ast.type === 'ReturnStatement') {
+          var applyCtxArgNode,
+              applyCtxNode;
+          if (extendedModeDetected === 'replace') {
+            applyCtxArgNode = ast.argument;
+          } else if (extendedModeDetected === 'extend') {
+            applyCtxArgNode = {
```

Comment:
Style! :) Missing semicolon, odd braces alignment. Please keep it the same as it is in the rest of the file.